QuickInfo should show the captures for lambdas and local functions

jcouv opened the pull request:
### Customer scenario
Performance-sensitive code should beware of captured variables when using lambdas. But it's not always easy to tell what variables are captured and will cause a closure to be created.
This feature adds a "Captures" section the QuickInfo bubble:

![image](https://user-images.githubusercontent.com/12466233/34902938-450b7cb2-f7db-11e7-9834-ae4ab0874e70.png)


![image](https://user-images.githubusercontent.com/12466233/34903013-be9b91c4-f7dc-11e7-88ed-1dd4e5fe58a6.png)


### Bugs this fixes
Fixes https://github.com/dotnet/roslyn/issues/23307

### Workarounds, if any
Figures out the captures mentally.

### Risk
### Performance impact

### Is this a regression from a previous update?
No.

alrz commented:
This is awesome. just two suggestions: it would be nice if we underline function name and lambda arrow as a visual clue if there is any capture (so you don't need to hover over any function/lambda to make sure it's not capturing, for example). also, it might make sense to group parameters and locals (if that's helpful at all).

jcouv commented:
I think I'll need Ali's PR (https://github.com/dotnet/roslyn/pull/23428) to get scenarios with multiple capture sets to work (`N(() =$$> { M(); }, () => { i++; });`).

sharwell commented:
@jcouv I'm a big fan of this and a few related items. I've marked this as a work in progress (in a way the bot understands) and will follow up in the original issue.

MaStr11 commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -587,6 +587,11 @@ public bool IsBindableToken(SyntaxToken token)
                 return true;
             }
 
+            if (token.IsKind(SyntaxKind.EqualsGreaterThanToken))
```
I did something comparable while adding support for QuickInfo for Linq query syntax [here](https://github.com/dotnet/roslyn/pull/23049/files#diff-1eda1d67f1b7acf2ea04b9af6d0e105aR594). I was wondering whether this might have unwanted side effects (In my case it is even worse, because it is a CommaToken). If you experience something unexpected it would be nice to get me informed.

jcouv replied:
@jason@jason-m.com Can you clarify the concern?
The method is only used in `GetSemanticInfoAtPositionAsync` which now supports finding a symbol when the position is in `=>`.
`GetSemanticInfoAtPositionAsync` itself is used in `GoToDefinition`, `Peek` and `SymbolFinder` (which is used in numerous places, like `FindUsages`, `FindReferences`, ...).

---
In reply to: [161932149](https://github.com/dotnet/roslyn/pull/23970#discussion_r161932149) [](ancestors = 161932149)

jasonmalinowski commented:
@jcouv I'll ignore the actual code for review until the WIP is removed, but this should probably target master and not be carried along with features/compiler. This is an IDE-only feature that can be shipped independent of new language features.

jcouv commented:
@jasonmalinowski This cannot go into master, as it depends on a new compiler API (`CapturesInside`) which was recently added to `features/compiler` for 15.7.

There are two test regressions, which I'm still investigating. When I ask for a semantic model for the some lambda syntaxes, I get back a `MethodBodySemanticModel`, which does not support flow analysis. I'm not sure why I'm getting that semantic model.

Also, @sharwell asked for IDE team to discuss this. It's probably better to hold-off review until design is approved.

CyrusNajmabadi commented:
```captures @this```?  Why the @?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -4761,6 +4761,124 @@ void M()
+        void local() { i++; this.M(); }
+    }
+}",
+                Captures($"\r\n{WorkspacesResources.Captures_colon} @this, i"));
```
@this?

CyrusNajmabadi commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,36 @@ private void AddExceptions(ISymbol symbol)
+                        parts.AddRange(ToMinimalDisplayParts(captured, s_formatForCaptures));
+                        first = false;
+                    }
+                    AddToGroup(SymbolDescriptionGroups.Captures, parts);
```
newlines after }  (above as well).

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -1799,5 +1804,10 @@ public void GetPartsOfCastExpression(SyntaxNode node, out SyntaxNode type, out S
+        public bool IdentifiesLambda(SyntaxToken token)
+        {
+            return token.IsKind(SyntaxKind.EqualsGreaterThanToken);
+        }
```
nit: use =>

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -341,6 +341,8 @@ internal interface ISyntaxFactsService : ILanguageService
 
         SyntaxToken? GetDeclarationIdentifierIfOverride(SyntaxToken token);
+
+        bool IdentifiesLambda(SyntaxToken token);
```
on the fence if this is the right location for this method.

jcouv replied:
I'm planning to leave as-is unless I hear some better suggestion ;-)

---
In reply to: [159457900](https://github.com/dotnet/roslyn/pull/23970#discussion_r159457900) [](ancestors = 159457900)

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -194,6 +194,19 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
+                allSymbols = semanticModel.GetSymbolInfo(bindableParent, cancellationToken)
+                                              .GetBestOrAllSymbols()
+                                              .WhereAsArray(s => !s.Equals(declaredSymbol))
+                                              .SelectAsArray(s => MapSymbol(s, type));
```
align dots.

jcouv commented on src/EditorFeatures/TestUtilities/QuickInfo/AbstractSemanticQuickInfoSourceTests.cs:
```diff
@@ -288,6 +288,14 @@ protected Action<object> Exceptions(string expectedText)
+                AssertTextAndClassifications(expectedText, expectedClassifications: null, actualContent: ((QuickInfoDisplayDeferredContent)content).CapturesText);
+            };
+        }
+
```
Felt harder to visually parse. Also, it's inconsistent with neighboring code.

CyrusNajmabadi replied:
> Also, it's inconsistent with neighboring code.

Fix the neighboring code :)

(But if you feel like it's actually harder to parse, that's a totally legitimate reason to not do it).

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs:
```diff
@@ -85,7 +85,7 @@ private static bool IsEscapable(SymbolDisplayPartKind kind)
         {
             var kind = SyntaxFacts.GetKeywordKind(identifier);
-            return kind == SyntaxKind.None
+            return kind == SyntaxKind.None || kind == SyntaxKind.ThisKeyword
```
This looks wrong.  this means we would not properly escape things in something like ```Foo(int @this)``` anymore.

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs:
```diff
@@ -4934,18 +4954,18 @@ public static void Main()
             Assert.Empty(flowAnalysis.CapturedInside);
             Assert.Empty(flowAnalysis.CapturedOutside);
-            Assert.Equal("MyClass @this", flowAnalysis.DataFlowsIn.Single().ToTestDisplayString());
+            Assert.Equal("MyClass this", flowAnalysis.DataFlowsIn.Single().ToTestDisplayString());
```
yup.  these are all bad changes.

jcouv replied:
No, I think those changes are correct. Added some explicit tests with "this" and "@this" to illustrate the point.

---
In reply to: [160014222](https://github.com/dotnet/roslyn/pull/23970#discussion_r160014222) [](ancestors = 160014222)

CyrusNajmabadi replied:
Gotcha!

CyrusNajmabadi commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,44 @@ private void AddExceptions(ISymbol symbol)
+                        }
+
+                        parts.AddRange(Space(count: 1));
+                        parts.AddRange(ToMinimalDisplayParts(captured, s_formatForCaptures));
```
I thinkt the better thing would be to use a format that asks to not escape keywords.  Alternatively, if 'this' is captured, just eplicitly write in the keyword yourself.

jcouv replied:
For the present situation, I'd rather fix the common machinery. "@this" and "this" should both be supported, depending on the symbol.

---
In reply to: [160014322](https://github.com/dotnet/roslyn/pull/23970#discussion_r160014322) [](ancestors = 160014322)

CyrusNajmabadi replied:
Works for me.

CyrusNajmabadi commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,44 @@ private void AddExceptions(ISymbol symbol)
+
+            private static readonly SymbolDisplayFormat s_formatForCaptures = SymbolDisplayFormat.MinimallyQualifiedFormat
+                .RemoveLocalOptions(SymbolDisplayLocalOptions.IncludeType)
+                .RemoveParameterOptions(SymbolDisplayParameterOptions.IncludeType);
```
Are you basically just writing the name of the local/parameter?  If you're removing the type, it's unclear why use symboldisplay or formats at all.

jcouv replied:
I think I addressed this elsewhere, but by understanding is that the symbol display handles not just strings, but also colorization. I'd rather rely on that machinery (and fix that shared code as appropriate).

---
In reply to: [160014346](https://github.com/dotnet/roslyn/pull/23970#discussion_r160014346) [](ancestors = 160014346)

CyrusNajmabadi replied:
Ah.  ok.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs:
```diff
@@ -82,12 +82,34 @@ private static bool IsEscapable(SymbolDisplayPartKind kind)
+                    break;
+            }
+
+            return escape ? $"@{identifier}" : identifier;
```
couldn't this be written: 

```c#
var kind = SyntaxFacts.GetKeywordKind(identifier);
return kind == SyntaxKind.None || (symbol as IParameterSymbol)?.IsThis == true
               ? identifier
               : $"@{identifier}";
```

All the flow control above seems overly confusing to me.

?

jcouv replied:
Layering prevents this. The extension is in Workspaces, but this code is in compiler.

---
In reply to: [161929932](https://github.com/dotnet/roslyn/pull/23970#discussion_r161929932) [](ancestors = 161929932)

jcouv commented on src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb:
```diff
@@ -2066,6 +2066,36 @@ End Namespace
+        <Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)>
+        <WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")>
+        Public Async Function TestQuickInfoCaptures() As Task
+            ' PROTOTYPE There is a problem with speculative semantic model scenario...
```
For some reason, VB hits the speculative semantic model case, whereas the corresponding C# tests didn't. Another mystery for another night...

jcouv commented:
Removed WIP tag since IDE design agreed with some feedback which I incorporated.
Note: region analysis isn't possible when speculating, so I adjusted the tests to account for that. When speculating the `Captures` section is left empty.

jcouv commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,44 @@ private void AddExceptions(ISymbol symbol)
+                var semanticModel = GetSemanticModel(syntax.SyntaxTree);
+                if (semanticModel.IsSpeculativeSemanticModel)
+                {
+                    // Speculative semantic models cannot do region analysis
```
@CyrusNajmabadi @dotnet/roslyn-ide  I'm not very familiar with speculation, so I'm not sure the impact of doing this (listing no captures when in speculative mode).

How often do we do QuickInfo in speculative mode? 
Should I list "Variables Captured: <unsaved document>" instead of removing the section?

CyrusNajmabadi replied:
Sorry, can you catch me up on things.  Is there a reason the speculative semantic model doesn't have capture information?

--

> How often do we do QuickInfo in speculative mode?

basically, we *always* try.  The idea of using speculative models is to get good perf for features that don't need things like exact positions of elements in source code.  Specifically, as a user is editing in a method, we'll commonly use a previously cached semantic model and then speculate against it using the current method body.  This means far less binding and other compiler overhead for many features that would not be able to tell the difference.

--

IMO, if we can't get speculative models to support providing capture info, then i think it might be acceptable to consider making quickinfo no longer use speculative models.  Quick info has far different perf concerns versus intellisense (teh major speculative benefactor).  People are typing fast, and intellisense needs to be as fast as it can to not slow that down.  Quick info is very different.  It's something a user needs to go and manually do with the mouse after typing.  That time spent interacting with the mouse means it's far more likely the normal semantic model will be ready, and the benefits of speculation will be minimal.

jasonmalinowski replied:
@jcouv Is there a deep reason that this API can't support speculation? Or is it mostly a paranoia check since we didn't want to test it?

gafter replied:
The speculative semantic model isn't modeling inserting the code at any particular position in the enclosing code, but merely somewhere within a given scope. You can point anywhere within a scope to speculate within that scope. We don't attempt to make sense of the speculative code as if it were somehow inserted into the flow of the program "at that point", as that's not how the API is supposed to work. If you want to imagine the enclosing code with some modification, you need to take that code and apply the modification you want. It would only make sense to do flow analysis in speculative code when you are speculating on a complete replacement for the method body.

jcouv replied:
Here's what I gathered from discussion with @gafter:

Flow analysis does not make sense when speculating.
Speculating is asking semantic model questions about a speculative snippet in the context of a given *scope* of existing code. Although the scope is specified by pointing into the existing code, speculation only reasons about the scope that position is in.
Within that scope, we don't know (and shouldn't assume) where the speculative snippet might be inserted, so flow analysis could not be done as the code (and therefore flow and regions) is undetermined.

jcouv replied:
From further testing in the IDE, it turns out I can hit the speculative case (shown below). Once I get that message, I'm not even sure how to get back to non-speculative case (this message seems very sticky 😞).
I think this feature will depend on removing speculation from QuickInfo as Cyrus suggested. I'll look over the weekend, but I may need some help with that.

![image](https://user-images.githubusercontent.com/12466233/35172564-04ff2f78-fd1d-11e7-93a1-0759b1297da0.png)

CyrusNajmabadi commented:
@gafter @jcouv that explanation makes sense.  thanks!

Is there enough built in telemetry to know if it would be a perf problem to stop using speculative semantic models for quickinfo?  My gut tells me it should be fine.  But having data to back that up would make the decision easier.

jcouv commented:
@CyrusNajmabadi I have no idea. I don't know how much the IDE relies on speculation.
I'll do some more manual testing in IDE to see how this change behaves.

CyrusNajmabadi commented:
> @CyrusNajmabadi I have no idea. I don't know how much the IDE relies on speculation.
I'll do some more manual testing in IDE to see how this change behaves.

IMO, speculation is really critical for things that are interacting with typing.  i.e. completion and sig-help.  For quick info, i think it's a lot less important.  I think we did it just because we could.  but it's not really critical there and we could totally avoid it.

@jasonmalinowski for his thoughts on this.

jasonmalinowski commented:
Agreeing with @CyrusNajmabadi; moving away from speculation wouldn't be a big worry for me. Did somebody dig through and see when we moved to speculation to see if there was a particular perf problem being addressed?

CyrusNajmabadi commented:
IIRC, i may have done it.  I think it was just done for symmetry, especially as we have the same MVC approach to those three features.

jasonmalinowski commented:
Also @jcouv I suspect your PR description and pictures need updating at this point as I think they're out of sync with the code?

jcouv commented:
@jasonmalinowski Updated the screenshots in description.

jcouv commented:
I've done some more manual validation in IDE and didn't notice anything unexpected. Basically, I'm not sure how or when speculation comes into play.
@dotnet/roslyn-ide for review. Thanks

CyrusNajmabadi commented:
>  Basically, I'm not sure how or when speculation comes into play.

usually it's very hard to observe.  It happens when typing fast and using some IDE features immediately after keystrokes.  In this case, we'll take a previous semantic model, and speculate the current nodes against it.  However, you need to be really fast (or the machine needs to be bogged down).  This is because so many features light up that we basically will always have a real semantic model that can be used in a very short period after any edit.  Where this is super valuable is completion lists.  That's because you have the scenario where an edit is made, and you *immediately* want accurate completion on a tree that is brand new to the compiler.  This is the worst-of-all-worlds for semantic information, so speculative models are critical.  Similarly for sig-help (though wrong or slow sig help doesn't interfere with typing the way wrong or slow completion can).

For quick info, there's virtually no point.  By the time you move your cursor to get any quick info, it's likely that the IDE has caught up and has a non-speculative model to work with.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/DeferredContent/QuickInfoDisplayDeferredContent.cs:
```diff
@@ -34,6 +36,7 @@ internal class QuickInfoDisplayDeferredContent : IDeferredQuickInfoContent
             AnonymousTypes = anonymousTypes;
             UsageText = usageText;
             ExceptionText = exceptionText;
+            CapturesText = capturesText;
```
FYI to both @JieCarolHu and @DustinCampbell as a collision warning for other quick info things happening.

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/AbstractQuickInfoProvider.cs:
```diff
@@ -102,7 +104,8 @@ private IDeferredQuickInfoContent CreateWarningGlyph()
             IList<TaggedText> usageText,
-            IList<TaggedText> exceptionText)
+            IList<TaggedText> exceptionText
+          )
```
uh, what?

jasonmalinowski commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,44 @@ private void AddExceptions(ISymbol symbol)
+                        }
+
+                        parts.AddRange(Space(count: 1));
+                        parts.AddRange(ToMinimalDisplayParts(captured, s_formatForCaptures));
```
What kind is being used here for 'this'/'Me'? Keyword (so we get colorization) or no?

jcouv replied:
The parts seem to have the right kinds (shown below). But it looks like parameters and locals are all rendered with the same color anyways (white in my theme).

![image](https://user-images.githubusercontent.com/12466233/35176033-9cdc6a8c-fd2b-11e7-9e47-bbd0c32722e8.png)
![image](https://user-images.githubusercontent.com/12466233/35176244-b677cf3a-fd2c-11e7-9249-be70027338eb.png)

jasonmalinowski replied:
I guess I'd expect the magic this parameter to be called a keyword, not a parameter. Our symbol model indeed calls it parameter; users don't.

jasonmalinowski commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -4761,6 +4752,141 @@ void M()
+        [WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")]
+        public async Task QuickInfoCapturesOnDelegate()
+        {
+            await TestWithoutSpeculativeAsync(@"
```
Why would some tests test with speculative and some without it? We're only doing one thing in the quick info driver, right?

jcouv replied:
The expected result when not using speculation is shown below, but the expected result when using speculation is empty.
I have not found a way (without extensive rewiring) to make a single test successfully check both.
So this test only checks the non-speculative case.
I can add test for the speculative case, but I'm not sure it's useful because it complicates the test machinery (need a tri-flag instead of bool) and it is trivial (empty result).

---
In reply to: [161964225](https://github.com/dotnet/roslyn/pull/23970#discussion_r161964225) [](ancestors = 161964225)

CyrusNajmabadi replied:
if you're no longer having quick-info use a speculative model (which i havne't verified is what you're actually doing), is there any need for the quick-info tests to do anything special anymore?

jcouv replied:
I've not been able to remove the usage of speculation yet. That's the main open issue that I'm working on.

jcouv replied:
I removed the calls to `GetSemanticModelForNodeAsync` from the QuickInfo provider. 
The IDE now behaves as I'd expect (the captures are populated even after I edit the source).

Thanks for the tips! 😃

jcouv commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +176,44 @@ private void AddExceptions(ISymbol symbol)
+                if (semanticModel.IsSpeculativeSemanticModel)
+                {
+                    // Speculative semantic models cannot do region analysis
+                    return;
```
I don't think so, but I'm not definitive.

---
In reply to: [161930944](https://github.com/dotnet/roslyn/pull/23970#discussion_r161930944) [](ancestors = 161930944)

CyrusNajmabadi commented:
https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/AbstractSemanticQuickInfoProvider.cs#L269-L275

Don't use GetSemanticModelForNodeAsync.  Just use GetSemanticModelAsync.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs:
```diff
@@ -194,17 +194,30 @@ private static ISymbol MapSymbol(ISymbol symbol, ITypeSymbol type)
                 declaredSymbol = null;
                 allSymbols = overriddenSymbol is null ? ImmutableArray<ISymbol>.Empty : ImmutableArray.Create(overriddenSymbol);
             }
+            else if (syntaxFacts.IdentifiesLambda(token))
```
I'm concerned here as well. GetSemanticInfo is used by several features (including Rename).  It's unclear ot me what the fallout of this would be.  I think i'd prefer that QuickInfo itself just have a fallback path for lambdas (since that's the scenario we're trying to fix up).

jcouv replied:
Thanks for the clarification.
I'll push a fix shortly. That also resolved the F1 regression on => so I'm happy :-)

---
In reply to: [162739605](https://github.com/dotnet/roslyn/pull/23970#discussion_r162739605) [](ancestors = 162739605)

CyrusNajmabadi commented on src/EditorFeatures/CSharp/QuickInfo/SemanticQuickInfoProvider.cs:
```diff
@@ -13,6 +13,14 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.QuickInfo
+        /// </summary>
+        protected override SyntaxNode GetBindableNodeForTokenIndicatingLambda(SyntaxToken token)
+        {
+            return token.Parent;
```
do you want to do any validation that token is the => token?

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/AbstractSemanticQuickInfoProvider.cs:
```diff
@@ -274,8 +283,18 @@ internal abstract partial class AbstractSemanticQuickInfoProvider : AbstractQuic
+            var syntaxFacts = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
+
+            ImmutableArray<ISymbol> symbols;
+            if (syntaxFacts.IdentifiesLambda(token))
```
yeah. i would just have IdentifiesLambda just pulled into this type as an abstract method.

jcouv replied:
Done :-)

CyrusNajmabadi commented on src/EditorFeatures/VisualBasic/QuickInfo/SemanticQuickInfoProvider.vb:
```diff
@@ -92,6 +92,13 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.QuickInfo
 
+        ''' <summary>
+        ''' Given a Sub or Function token for a lambda, returns the syntax for the whole lambda
+        ''' </summary>
```
do we have tests for single/multiline sub/function lambdas?

jcouv replied:
Yes, I have all four cases

jcouv commented on src/VisualStudio/CSharp/Test/F1Help/F1HelpTests.cs:
```diff
@@ -501,7 +501,7 @@ void M()
         });
     }
-}", "=>_CSharpKeyword");
+}", "C.lambda expression");
```
Good catch.
By moving the logic to get the lambda symbol down into the QuickInfo provider, this regression got fixed.

jcouv commented:
Fixed a couple of test regressions. 
One case required a small code change: when invoking symbol completion on a local, we will build a description for that local using speculation, but we'll only display the "main description" portion of that description (not the captures).

@dotnet/roslyn-ide for review. Thanks

jcouv commented:
@dotnet/roslyn-ide for review. All the feedback so far was addressed. Thanks

CyrusNajmabadi commented on src/EditorFeatures/CSharp/QuickInfo/SemanticQuickInfoProvider.cs:
```diff
@@ -13,6 +14,31 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.QuickInfo
+                && token.Parent.IsKind(SyntaxKind.ParenthesizedLambdaExpression, SyntaxKind.SimpleLambdaExpression))
+            {
+                // () =>
+            }
```
not sure how i feel about this style of no-body then falling out to common code :)

CyrusNajmabadi commented on src/EditorFeatures/VisualBasic/QuickInfo/SemanticQuickInfoProvider.vb:
```diff
@@ -92,6 +93,19 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.QuickInfo
+        ''' If the token is a 'Sub' or 'Function' in a lambda, returns the syntax for the whole lambda
+        ''' </summary>
+        Protected Overrides Function GetBindableNodeForTokenIndicatingLambda(token As SyntaxToken, <Out> ByRef found As SyntaxNode) As Boolean
+            If token.IsKind(SyntaxKind.SubKeyword, SyntaxKind.FunctionKeyword) AndAlso token.Parent.IsKind(SyntaxKind.SubLambdaHeader, SyntaxKind.FunctionLambdaHeader) Then
```
to verify: this works for single and multi-line VB lambdas?

jcouv replied:
Yes. Here's what the syntax tree looks like (for Sub):

[Multi-line Sub](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBsQDdhJiaMA+AwitFAASECwAUGXWQMoCuwZAsgBQAeZAguQEkAdvADmiAJQ16MsgBEAlmDIQ+5BgE9YcMADpeAYxgKA9kLIBeRiw5Tashzys8A1GQCM0h2QCiQgCbWwF50foHMwdRhFCRQUEA=)
![image](https://user-images.githubusercontent.com/12466233/35416433-ec0d6d8e-01dd-11e8-92e5-062178443d25.png)

[Single-line Sub](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBsQDdhJiaMA+AwitFAASECwAUGXWQMoCuwZAsgBQAeZAguQEkAdvADmiAJQ16MsgBEAlmDIQ+5BgE9YcMADpeAYxgKA9kLIBeRiw4SyPKzwDUZAIw1p9AKJCAJteAaH39iUiA=)

![image](https://user-images.githubusercontent.com/12466233/35416442-f4ca446a-01dd-11e8-98e8-cd3b2594e3a5.png)

Also, I have tests for both multi-line and single-line:
```VB

        <Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)>
        <WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")>
        Public Async Function TestQuickInfoCaptures() As Task
            Await TestAsync("
Class C
    Sub M(x As Integer)
        Dim a As System.Action = Sub$$()
            x = x + 1
        End Sub
    End Sub
End Class
",
                Captures($"{vbCrLf}{WorkspacesResources.Variables_captured_colon} x"))
        End Function

        <Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)>
        <WorkItem(23307, "https://github.com/dotnet/roslyn/issues/23307")>
        Public Async Function TestQuickInfoCaptures2() As Task
            Await TestAsync("
Class C
    Sub M(x As Integer)
        Dim a As System.Action = S$$ub() x = x + 1
    End Sub
End Class
",
                Captures($"{vbCrLf}{WorkspacesResources.Variables_captured_colon} x"))
        End Function
```

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:
```diff
@@ -26,11 +26,11 @@ public class SemanticQuickInfoSourceTests : AbstractSemanticQuickInfoSourceTests
         }
 
-        private async Task TestWithOptionsAsync(TestWorkspace workspace, params Action<object>[] expectedResults)
+        private async Task TestWithOptionsAsync(TestWorkspace workspace, bool skipSpeculative, params Action<object>[] expectedResults)
```
i'm a little confused as to why we would still need this.  Since quick-info is no longer using any speculative semantic models, there should be no effect of trying with/without speculation.  Indeed, i would be fine if the CanUseSpeculativeSemanticModelAsync part of the test was removed entirely (though keeping it would be fine.)

jcouv replied:
Sorry about that. It was a left-over I should have cleaned up. Thanks

CyrusNajmabadi commented on src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:
```diff
@@ -175,6 +177,54 @@ private void AddExceptions(ISymbol symbol)
+                var semanticModel = GetSemanticModel(syntax.SyntaxTree);
+                if(semanticModel.IsSpeculativeSemanticModel)
+                {
+                    // In the context of symbol completion, it is possible we'll make a description for a symbol while speculating (but it won't be displayed)
```
lolwut?  we make a description, but don't display it?  

Also, if that's the case, why add the group at all?  why not just insta bail?

Finally, if you are going to check IsSpeculativeSemanticModel, it's worth calling that that's because data flow doens't work with SpeculativeSemanticModels, so you need to avoid the code below this check.

jcouv replied:
I'll clarify the comment. When you complete a local symbol, we make a description (with many parts: main description, exceptions, captures, etc), but only the main description seems to be displayed.
The reason I put captures here anyways, is in case I'm wrong and missed some scenario. Captures listing "?" is less ambiguous/misleading than displaying no captures at all.

Thanks!

jasonmalinowski commented on src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/AbstractSemanticQuickInfoProvider.cs:
```diff
@@ -266,16 +273,28 @@ internal abstract partial class AbstractSemanticQuickInfoProvider : AbstractQuic
             return CreateDocumentationCommentDeferredContent(null);
         }
 
+        protected abstract bool GetBindableNodeForTokenIndicatingLambda(SyntaxToken token, out SyntaxNode found);
```
To check: this isn't being used by TypeScript or F# that this is a breaking change to derivers? I don't see _how_ they could , but want to confirm.

jcouv replied:
I checked with F# folks (Will and Brett) and they do not reference `AbstractSemanticQuickInfoProvider` from Roslyn (they only depend on some interfaces for quick info provider.

Same thing with TypeScript (Mine Starks), which only depends on interfaces, but not this abstract type.

jcouv commented:
@jasonmalinowski Good catch. I fixed the compiler tests to check part kinds too. `this` and `Me` are now keywords.
This does result in nicer UI, and the code is also simpler that way.
![image](https://user-images.githubusercontent.com/12466233/35465408-8c13deda-02b0-11e8-9ef8-f4ebeb12773d.png)
Thanks!

@dotnet/roslyn-compiler for review. The compiler portion of this PR is a small change with how we display parts for `this` versus `@this` (and `Me` versus `[Me]`).

jasonmalinowski commented:
> This does result in nicer UI, and the code is also simpler that way.

Love it when that happens!

jcouv commented:
@VSadov raised a good question with nested scopes. Here's what I observe with nesting:

The inner local looks good:
![image](https://user-images.githubusercontent.com/12466233/35465873-21157e32-02b4-11e8-9e75-fa764d46a72b.png)

But the outer local says that it captures `j`, which is weird...
![image](https://user-images.githubusercontent.com/12466233/35465882-354a6f2a-02b4-11e8-9a4e-06f0af920176.png)
